### PR TITLE
[macOS Monterey and Big Sur] Enable AVIF image decoding using libavif and dav1d

### PR DESCRIPTION
**Pull request by litherum**
#### 5300e66a81df105c2dd2aefe87f42e95a27c027d
<pre>
[macOS Monterey and Big Sur] Enable AVIF image decoding using libavif and dav1d
<a href="https://bugs.webkit.org/show_bug.cgi?id=246925">https://bugs.webkit.org/show_bug.cgi?id=246925</a>
rdar://101481679

Reviewed by Said Abou-Hallawa.

We checked in libavif and dav1d sources in 494a012c9820b947bdad5b6b5ea95df04e5bb1fe.

This patch hooks up image decoding to use them, on macOS Monterey and Big Sur. It does
so by using the same glue code that non-Apple platforms use: ScalableImageDecoder and
AVIFImageDecoder. This patch includes those files in Sources.txt for all ports, and
refactors ScalableImageDecoder on Cocoa platforms to only support AVIF, as per
<a href="https://webkit.slack.com/archives/CU64U6FDW/p1666556573569359?thread_ts=1666555605.109619&amp">https://webkit.slack.com/archives/CU64U6FDW/p1666556573569359?thread_ts=1666555605.109619&amp</a>;cid=CU64U6FDW.

This patch re-uses the same USE(AVIF) flag that the non-Apple ports have been using.
There are currently 2 AVIF flags:
- USE(AVIF) indicates that libavif will be used to decode AVIF images, using the
      ScalableImageDecoder infrastructure. This is enabled on the non-Apple ports,
      and on macOS Monterey and Big Sur.
- HAVE(AVIF) indicates that the platform can decode AVIF images directly. This is enabled
      on macOS Ventura and later (and other Cocoa ports).
These flags are mutually exclusive; no port has both enabled at the same time. The Apple
Windows port has neither of these flags enabled.

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/libavif/Configurations/libavif.xcconfig:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageDecoder.cpp:
(WebCore::ImageDecoder::create):
* Source/WebCore/platform/graphics/cg/ImageBackingStoreCG.cpp: Copied from Source/WebCore/platform/image-decoders/avif/AVIFImageDecoder.h.
(WebCore::ImageBackingStore::image const):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::decodeUTI const):
(WebCore::ImageDecoderCG::decodeUTI):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
(WebCore::ScalableImageDecoder::create):
* Source/WebCore/platform/image-decoders/avif/AVIFImageDecoder.h:
* Source/WebCore/platform/image-decoders/avif/AVIFImageReader.cpp:
(WebCore::AVIFImageReader::decodeFrame):

Canonical link: <a href="https://commits.webkit.org/255984@main">https://commits.webkit.org/255984@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/eb1134f12bf39d5cbac0b12713b9fb2be0f66d27

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94219 "Passed style check") | [loading 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3411 "Build is being retried. Recent messages:validate-change running; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/25015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/103901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164177 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98220 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3438 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/31625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86573 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99875 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80631 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/25015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85457 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38007 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/25015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80594 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35891 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/25015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27728 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4139 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39769 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83258 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41712 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/25015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18798 "Passed tests") |

### CODE REVIEW

File: Source/WebCore/platform/graphics/ImageDecoder.cpp [hunk 1]
```diff
@@ -26,12 +26,11 @@
 #include "config.h"
 #include "ImageDecoder.h"
 
+#include "ScalableImageDecoder.h"
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/ImageDecoder.cpp]: Why is this included for Apple ports when USE(AVIF) is false?

File: Source/WebCore/platform/graphics/ImageDecoder.cpp [hunk 2]
```diff
@@ -104,6 +103,9 @@ RefPtr<ImageDecoder> ImageDecoder::create(FragmentedSharedBuffer& data, const St
 #endif
 
 #if USE(CG)
+    // ScalableImageDecoder is used on CG ports for some specific image formats which the platform doesn't support directly.
+    if (auto imageDecoder = ScalableImageDecoder::create(data, alphaOption, gammaAndColorProfileOption))
+        return imageDecoder;
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/ImageDecoder.cpp:112]: This is a little bit confusing. It makes the code looks like we are giving the open source decoders priority over the system decoders. I had to check ScalableImageDecoder::create() to know that it will return null if USE(AVIF) is false.

I think it will be better if we surround all the AVIF supporting code on down levels by #if USE(AVIF). So we can delete it all easily once we stop supporting macOS Monetary.

File: Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp [hunk 3]
```diff
@@ -68,72 +75,85 @@ static unsigned copyFromSharedBuffer(char* buffer, unsigned bufferLength, const
...
+static bool matchesCURSignature(char* contents)
+{
+    return !memcmp(contents, "\x00\x00\x02\x00", 4);
+}
+#endif
+
 #if USE(AVIF)
-bool matchesAVIFSignature(char* contents)
+#if USE(CG)
+static bool matchesAVIFSignature(char*, FragmentedSharedBuffer& data)
```
Comment (shallawa) [inline Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp]: I am not sure what the coding style we follow here is. But I think most of the cases, we author a single function and put the directives inside:
```
static bool matchesAVIFSignature(char*, FragmentedSharedBuffer& data)
{
#if USE(CG)
#else
#endif
}
```

File: Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp [hunk 4]
```diff
@@ -68,72 +75,85 @@ static unsigned copyFromSharedBuffer(char* buffer, unsigned bufferLength, const
...
+{
+    auto sharedBuffer = data.makeContiguous();
+    auto cfData = sharedBuffer->createCFData();
+    auto imageSource = adoptCF(CGImageSourceCreateWithData(cfData.get(), nullptr));
```
Comment (shallawa) [inline Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:117]: I think this means we will create two ImageSources for every non AVIF image: one here and one in ImageDecoderCG. How does affect PLT for example?

**Review by shallawa — APPROVED** [review]
Should not all/some of the skipped AVIF tests be un-skipped with this change?
```
fast/images/avif-image-decoding.html [ Skip ]
fast/images/avif-as-image.html [ Skip ]
fast/images/avif-heif-container-as-image.html [ Skip ]
fast/images/animated-avif.html [ Skip ]
http/tests/images/avif-partial-load-crash.html [ Skip ]
```

File: Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp (hunk 4 above)
Comment (litherum) [reply Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:116]: Good point. I'll run PLT.

Comment (litherum) [reply Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:117]: Oh, following your advice above, I moved this code behind `USE(AVIF)` which isn't enabled on Ventura, which is where we test PLT. So I think your advice above solves both of these concerns 👍
